MINIFICPP-1036 - libarchive extension improvements

bakaid opened the pull request:
This commit
 - transfers the libarchive compilation to an ExternalProject
 - adds bundled liblzma and bzip2
 - makes the libarchive extension compile on Windows
 - adds the "Encapsulate in TAR" Property to CompressContent
 - adds raw GZIP compression support to CompressContent

Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

arpadboda commented on libminifi/src/io/ZlibStream.cpp:
```diff
@@ -0,0 +1,202 @@
+}
+
+void ZlibCompressStream::closeStream() {
+  if (state_ == ZlibStreamState::INITIALIZED) {
```
In case error happened during write and close is called, what happens? Or we don't expect it to happen at all?

bakaid replied:
It's perfectly okay to call `closeStream` if an error happens, and we will do nothing: we will keep the error state as it was.

arpadboda commented on libminifi/src/io/ZlibStream.cpp:
```diff
@@ -0,0 +1,202 @@
+    ret = inflate(&strm_, Z_NO_FLUSH);
+    if (ret == Z_STREAM_ERROR ||
+        ret == Z_NEED_DICT ||
+        ret == Z_DATA_ERROR ||
```
Bleh, is this API designed so badly?

I would expect a success case or something, but handling some errors that look like a specific set of errors is weird.

bakaid replied:
This is how it is designed, see the example usage: https://zlib.net/zlib_how.html

arpadboda commented on libminifi/src/io/ZlibStream.cpp:
```diff
@@ -0,0 +1,202 @@
+}
+
+ZlibDecompressStream::~ZlibDecompressStream() {
+  if (state_ != ZlibStreamState::UNINITIALIZED) {
```
The same here, uninited state can't happen, although I wonder if inflateEnd should be called in case decompress failed and we have an error state.

bakaid replied:
inflateEnd frees the resources allocated by zlib, so if we called inflateInit and it succeeded we must call inflateEnd, even if there as an error.

arpadboda replied:
Okay, thanks, that makes sense.

szaszm replied:
Would it make sense to make `state_ != UNINITIALIZED` a postcondition of the ctor and remove the UNINITIALIZED state? It seems to be already the case, so we could remove all possibility of related dead code by removing the enum value. We don't need to prepare for any usage of an object, including destruction, during object creation.

bakaid replied:
Not if you ever want to make this object movable: in that case you would have to set the moved-from object to UNINITIALIZED to prevent it from freeing the z_stream.

arpadboda commented on libminifi/test/archive-tests/CompressContentTests.cpp:
```diff
@@ -932,3 +938,107 @@ TEST_CASE("DecompressFileXYLZMA", "[compressfiletest8]") {
+  // Check decompressed file
+  std::ifstream decompressed(decompressed_file, std::ios::in | std::ios::binary);
+  std::string decompressed_content((std::istreambuf_iterator<char>(decompressed)), std::istreambuf_iterator<char>());
+  REQUIRE(content == decompressed_content);
```
Nice test!

bakaid replied:
Thanks!

arpadboda commented on libminifi/test/unit/ZlibStreamTests.cpp:
```diff
@@ -0,0 +1,105 @@
+  SECTION("Simple content in two writes") {
+    REQUIRE(-1 != compressStream.write(reinterpret_cast<uint8_t*>(const_cast<char*>("foo")), strlen("foo")));
+    original += "foo";
+    REQUIRE(-1 != compressStream.write(reinterpret_cast<uint8_t*>(const_cast<char*>("bar")), strlen("bar")));
```
I think the return values should match the length of the string. 
Or in case they return the number of compressed bytes written, let's check for grearter than zero.

bakaid replied:
It must return the number of bytes consumed, will change to an exact check.

arpadboda commented on thirdparty/bzip2/bzip2.patch:
```diff
@@ -0,0 +1,31 @@
+diff -rupN orig/CMakeLists.txt patched/CMakeLists.txt
```
A justification for this file (in a comment or sg) would be nice.

bakaid replied:
bzip2 only provides a Makefile which is not very usable on Windows, and it is a pretty simple project so I created a CMake file to compile it on all platforms.

arpadboda commented on thirdparty/liblzma/liblzma.patch:
```diff
@@ -0,0 +1,118 @@
++
++project(liblzma)
++
++set(SOURCES src/common/tuklib_cpucores.c
```
Do we patch to only compile the subset of files/features we need?

bakaid replied:
We can't run the configure script on Windows, and the only thing lzma provides for Windows are msbuild files which do not integrate well into our CMake build system, so I wrote a windows-specific CMakeLists.txt for the project.

bakaid commented on libminifi/src/io/ZlibStream.cpp:
```diff
@@ -0,0 +1,202 @@
+}
+
+ZlibCompressStream::~ZlibCompressStream() {
+  if (state_ != ZlibStreamState::UNINITIALIZED) {
```
You are right, but I think it is better readable this way.

bakaid commented on extensions/libarchive/CompressContent.cpp:
```diff
@@ -39,6 +39,11 @@ core::Property CompressContent::CompressLevel("Compression Level", "The compress
 core::Property CompressContent::CompressMode("Mode", "Indicates whether the processor should compress content or decompress content.", MODE_COMPRESS);
 core::Property CompressContent::CompressFormat("Compression Format", "The compression format to use.", COMPRESSION_FORMAT_ATTRIBUTE);
 core::Property CompressContent::UpdateFileName("Update Filename", "Determines if filename extension need to be updated", "false");
+core::Property CompressContent::EncapsulateInTar("Encapsulate in TAR", "If true, on compression the FlowFile is added to a TAR archive and then compressed, "
```
I completely agree, but because all other Properties used the older API, I didn't want this to stand out.
However, after using this processor through C2, I am convinced that converting all Properties to use validators in this processor will worth the (not significant) effort, so I'll do that.

bakaid commented on extensions/libarchive/CompressContent.h:
```diff
@@ -341,6 +346,73 @@ class CompressContent: public core::Processor {
+            } else if (ret == 0) {
+              break;
+            } else {
+              if (outputStream_->writeData(buffer.data(), ret) != ret) {
+                return -1;
+              }
+              read_size += ret;
+            }
+          }
+          outputStream_->closeStream();
```
The generic answer is that our Stream API is not very highly (or properly) defined and whatever `closeStream` does, or when it should be called is implementation-defined.
In this particular case we have to call `closeStream` in the successful case, because that's how we signal the end of the input data and force zlib to flush its buffers and write out the end of the compressed stream.

bakaid commented on libminifi/include/io/ZlibStream.h:
```diff
@@ -0,0 +1,105 @@
+class ZlibCompressStream : public ZlibBaseStream {
+ public:
+  ZlibCompressStream(bool gzip = true, int level = Z_DEFAULT_COMPRESSION);
+  ZlibCompressStream(DataStream* other, bool gzip = true, int level = Z_DEFAULT_COMPRESSION);
```
Fair point. I'll replace the `bool gzip` with an enum class to the effect of `ZlibFormat format`. A pointer is not converted implicitly to an enum class, and it will also improve readability.

szaszm replied:
Could you also mark both ctors as `explicit`? Since this is new code, it doesn't break API compatibility.

bakaid replied:
I could, although I don't really see the point in this case.

szaszm replied:
To prevent implicit conversion from `bool`/`ZlibFormat` and `DataStream*`, as suggested in [C.46: By default, declare single-argument constructors explicit](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-explicit). (The guideline doesn't, but the reason covers all constructors that can be called with a single argument, i.e. including ctors with more parameters where all N>1-th parameter has a default argument.)

Or if you intended the implicit conversion to be part of the API, I'm also fine with an explanation and possibly an example of intended usage added in a code comment.

bakaid replied:
Fair, will mark them explicit.

bakaid commented on extensions/libarchive/CompressContent.h:
```diff
@@ -341,6 +346,73 @@ class CompressContent: public core::Processor {
+
+    ~GzipWriteCallback() override = default;
+
+    std::shared_ptr<logging::Logger> logger_;
+    std::string compress_mode_;
+    int64_t compress_level_;
+    std::shared_ptr<core::FlowFile> flow_;
+    std::shared_ptr<core::ProcessSession> session_;
+    bool success_{false};
```
Changed.

bakaid commented on extensions/libarchive/CompressContent.h:
```diff
@@ -341,6 +346,73 @@ class CompressContent: public core::Processor {
+    {
+    }
+
+    ~GzipWriteCallback() override = default;
```
Force of habit, removed.

bakaid commented on extensions/libarchive/CompressContent.cpp:
```diff
@@ -74,7 +80,13 @@ void CompressContent::onSchedule(core::ProcessContext *context, core::ProcessSes
   }
-  logger_->log_info("Compress Content: Mode [%s] Format [%s] Level [%d] UpdateFileName [%d]", compressMode_, compressFormat_, compressLevel_, updateFileName_);
+  value = "";
+  if (context->getProperty(EncapsulateInTar.getName(), value) && !value.empty()) {
```
Done.

bakaid commented on extensions/libarchive/ArchiveCommon.h:
```diff
@@ -0,0 +1,29 @@
+#ifdef WIN32
+#define mode_t uint16_t
+#define uid_t int64_t
+#define gid_t int64_t
```
Not sure, I think I was distracted by some third party header mess when doing this, typedef is the nice solution here, you are right. Changed.

bakaid commented:
@arpadboda @szaszm Addressed review comments, it also builds for me on Travis now (on Xenial for some unfathomable reason some constexpr string literals in the sftp extension started to get ODR-used and therefore lacked storage and build failed with a link error - even though I haven't even touched the extension. Anyway, added storage for them in the last commit and it builds properly now.)

szaszm commented on libminifi/test/archive-tests/CompressContentTests.cpp:
```diff
@@ -932,3 +938,107 @@ TEST_CASE("DecompressFileXYLZMA", "[compressfiletest8]") {
+    content = content_ss.str();
+  }
+
+  std::fstream file;
+  file.open(src_file, std::ios::out);
+  file << content;
+  file.close();
```
Could be a oneliner: `std::ofstream{src_file} << content;`. Just an idea, I don't mind if it stays like this.

bakaid replied:
This is a test, I copied this from somewhere, I could not really be bothered whether its one line or four.

bakaid commented on extensions/libarchive/CompressContent.cpp:
```diff
@@ -35,10 +35,32 @@ namespace nifi {
+core::Property CompressContent::CompressFormat(
+    core::PropertyBuilder::createProperty("Compression Format")->withDescription("The compression format to use.")
+        ->isRequired(false)
+        ->withAllowableValues<std::string>({
+          COMPRESSION_FORMAT_ATTRIBUTE,
+          COMPRESSION_FORMAT_GZIP,
+          COMPRESSION_FORMAT_BZIP2,
+          COMPRESSION_FORMAT_XZ_LZMA2,
```
Could you elaborate? Any YAML config that didn't use values from this enum was erroneous and resulted in an error on onTrigger. How this change would break those configurations further?

szaszm replied:
In that case, I was wrong.

bakaid commented on libminifi/include/io/ZlibStream.h:
```diff
@@ -0,0 +1,110 @@
+class ZlibBaseStream : public BaseStream {
+ public:
+  ZlibBaseStream();
+  ZlibBaseStream(DataStream* other);
+
+  ZlibBaseStream(const ZlibBaseStream&) = delete;
+  ZlibBaseStream& operator=(const ZlibBaseStream&) = delete;
+  ZlibBaseStream(ZlibBaseStream&& other) = delete;
+  ZlibBaseStream& operator=(ZlibBaseStream&& other) = delete;
+
+  ~ZlibBaseStream() override = default;
```
Done.

bakaid commented on libminifi/include/io/ZlibStream.h:
```diff
@@ -0,0 +1,110 @@
+
+class ZlibBaseStream : public BaseStream {
+ public:
+  ZlibBaseStream();
+  ZlibBaseStream(DataStream* other);
+
+  ZlibBaseStream(const ZlibBaseStream&) = delete;
+  ZlibBaseStream& operator=(const ZlibBaseStream&) = delete;
+  ZlibBaseStream(ZlibBaseStream&& other) = delete;
+  ZlibBaseStream& operator=(ZlibBaseStream&& other) = delete;
```
Done.

bakaid commented:
Addressed new review comments.
I would appreciate if we could move this PR along.
